Introduce eql search status API

mayya-sharipova opened the pull request:
Introduce eql search status API,
that reports the status of eql stored or async search.

GET _eql/search/status/<id>

The API is restricted to the monitoring_user role.

For a running eql search, a response has the following format:

```js
{
  "id" : <id>,
  "is_running" : true,
  "is_partial" : true,
  "start_time_in_millis" : 1611690235000,
  "expiration_time_in_millis" : 1611690295000
}
```
For a completed eql search, a response has the following format:
```js
{
  "id" : <id>,
  "is_running" : false,
  "is_partial" : false,
  "expiration_time_in_millis" : 1611690295000,
  "completion_status" : 200
}
```
Closes #66955

mayya-sharipova committed:
Introduce eql search status API

Introduce eql search status API,
that reports the status of eql stored or async search.

GET _eql/search/status/<id>

The API is restricted to the monitoring_user role.

For a running eql search, a response has the following format:

{
  "id" : <id>,
  "is_running" : true,
  "is_partial" : true,
  "start_time_in_millis" : 1611690235000,
  "expiration_time_in_millis" : 1611690295000
}

For a completed eql search, a response has the following format:

{
  "id" : <id>,
  "is_running" : false,
  "is_partial" : false,
  "expiration_time_in_millis" : 1611690295000,
  "completion_status" : 200
}

Closes #66955

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

mayya-sharipova commented:
@lizozom Unfortunately for eql searches, we could not the same response format as for async searches, as we don't store shards progress for eql searches. I hope the format presented above works for you.

imotov commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlStatusResponse.java:
```diff
@@ -0,0 +1,171 @@
+/**
+ * A response for eql search status request
+ */
+public class EqlStatusResponse extends ActionResponse implements SearchStatusResponse, StatusToXContentObject {
```
It might be good idea to add a AbstractSerializingTestCase-based test for this class to unit test serialization.

mayya-sharipova replied:
@imotov thank for the feedback, addressed in 4321efb4b62fc5a0fafdfc9c3e613f93a57a73bb

imotov commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/SearchStatusResponse.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * An interface for status response of the stored or running async search
+ */
+public interface SearchStatusResponse {
```
There is quite a bit of code repetition between AsyncStatusResponse and EqlStatusResponse, and between TransportEqlAsyncGetStatusAction and TransportGetAsyncStatusAction, any reason for not moving more shared things into SearchStatusResponse and AsyncTaskIndexService respectively?

mayya-sharipova replied:
@imotov  Thank for the feedback, addressed in 4321efb4b62fc5a0fafdfc9c3e613f93a57a73bb.

I was not able to combine AsyncStatusResponse and EqlStatusResponse together because it would complicate the wire  serialization/deserialization code.  As we already have `AsyncStatusResponse` form 7.11, I did not want to change that part of the code.

imotov reviewed (commented):
Looks good to me in general. Left a couple of comments. I think it would be great if @costin could also take a look.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlAsyncGetStatusAction.java:
```diff
@@ -0,0 +1,114 @@
+    protected void doExecute(Task task, GetAsyncStatusRequest request, ActionListener<EqlStatusResponse> listener) {
+        AsyncExecutionId searchId = AsyncExecutionId.decode(request.getId());
+        DiscoveryNode node = clusterService.state().nodes().get(searchId.getTaskId().getNodeId());
+        if (node == null || Objects.equals(node, clusterService.localNode())) {
```
More like a question, than an issue: why using `clusterService.localNode()` and not `clusterService.state().nodes().localNode()`?

mayya-sharipova replied:
@astefan  Thanks for checking. Apparently `clusterService.localNode()` throws an exception if `state().getNodes().getLocalNode()` is null; while `clusterService.state().nodes().localNode()` just returns `null` in this case.  I guess  in this situation, we don't want to throw an exception, and want to still proceed, but I wonder how it is possible for a local node to become `null`?

mayya-sharipova replied:
@astefan Thank for the feedback, addressed in 4321efb4b62fc5a0fafdfc9c3e613f93a57a73bb

mayya-sharipova commented:
@imotov @astefan Thanks for reviewing the PR. I will study and address your comments. 
One questions I have is that the Kibana team is interested to show the progress of EQL async search task, but it looks like currently EQL doesn't store partial results.  Is there a plan to store partial results in a similar way how async search stores them?

costin commented:
LGTM to me from my end though I'm not familiar with the async code so I'm non-binding :)
Regarding the EQL results, and this should be a separate PR, how can the partial information be exposed? The completed sequences (or in case of events query searches) are kept based on their doc ids and their sources retrieved at the end (see https://github.com/elastic/elasticsearch/blob/ef26b53f9359024a87225cd182a96248eebb6366/x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java#L250).
However it should be trivial to signal a new match during the process.

mayya-sharipova commented:
@costin  Thank you for checking. 

> Regarding the EQL results, and this should be a separate PR, how can the partial information be exposed? 

For a partial information for a status report, we don't need partial results, Kibana just needs progress indication about the number of the completed shards so far and total shards.
But overall, I was wondering if the EQL team has plans to store partial results or is that something that doesn't make sense for EQL and is not planned?

mark-vieira commented:
@elasticmachine update branch

mayya-sharipova committed:
Address feedback

- Add serializing tests for EqlStatusResponse and AsyncStatusResponse
- Move shared code of TransportEqlAsyncGetStatusAction and
  and TransportGetAsyncStatusAction to AsyncTaskIndexService
- Use clusterService.state().getNodes().getLocalNode()
  to get a local node

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into eql-async-search-status

mayya-sharipova committed:
Update license

Also remove skip section in yml tests, as these tests
are not used for BWC testings

mayya-sharipova committed:
Add eql/async/status to non-operator actions

imotov commented on x-pack/plugin/eql/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/eql/10_basic.yml:
```diff
@@ -158,3 +158,55 @@ setup:
+      eql.get_status:
+        id: $id
+  - match: { id: $id }
+  - match: { is_running: true }
+  - match: { is_partial: true }
```
These two checks will probably fail occasionally due to a race condition with task completion.

mayya-sharipova replied:
@imotov  Thanks! Good  feedback! Indeed,  we are not certain if the task is completed or not by this time!  I changed the test in 4eca9a0b6fe029fc245cd1c7612717fac1f13ba3 to test for less.

imotov reviewed (approved):
Thanks for the modifications! Left a small comment about a potential race condition in the integration test. Otherwise LGTM.

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into eql-async-search-status

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into eql-async-search-status

mayya-sharipova committed:
Client with origin

mayya-sharipova committed:
Adjust test

mayya-sharipova committed:
Address style
